### ISSUE TITLE
A better way to install stackrox operator without OLM

### ISSUE THREAD

**Issue by NissesSenap** [stackrox/stackrox#1401]
Today stackrox you have a great way of installing the stackrox operator through OLM.
But when trying to install the operator through kubernetes it's not as easy.

As you mentioned on the first community meeting everyone knows how to use helm charts to it probably would be a good way to solve this problem.

The bad thing is of course that we would have to maintain yet another helm chart.

Stackrox-chart is already taken by redhat-cop and is a helm chart to install the operator through OLM in a gitops scenario:
https://artifacthub.io/packages/helm/redhat-cop/stackrox-chart
https://github.com/redhat-cop/helm-charts/tree/master/charts/stackrox

But a good name could be as simple as stackrox-operator.

Why?
To be able to grow the community outside of Openshift we need to provide the kubernetes users with a good way of installing the operator.

**Comment by md2119** [comment]
@NissesSenap, thank you for contributing to the StackRox community!
I will direct your feedback to the appropriate engineering team. We'll update this issue as soon as we can provide more insight.

**Comment by SimonBaeumer** [comment]
Hey @NissesSenap,
glad to see all your feature requests. Thanks!
Supporting our operator on non-OpenShift is on our roadmap but not scheduled yet and needs some more investigations, i.e. installing operator via Helm and auto-upgrading outside of OLM.

Best,
Simon

**Comment by NissesSenap** [comment]
Fully understand @SimonBaeumer. My plan is to try to join the next community meeting. If possible it would be nice to have a short discussion about possible ways forward.

I could definitely help out with some basic CI/CD setup and write the initial helm chart. The question in my mind is how to manage it in a good way long term.

**Comment by SimonBaeumer** [comment]
> Fully understand @SimonBaeumer. My plan is to try to join the next community meeting. If possible it would be nice to have a short discussion about possible ways forward.
> 
> I could definitely help out with some basic CI/CD setup and write the initial helm chart. The question in my mind is how to manage it in a good way long term.

@NissesSenap That's great to hear, thank you! I'll join the community meeting (10th May @ 18:00 CET).

**Comment by NissesSenap** [comment]
Had a chat about this during the community meeting.
We will create a separate repo for the helm chart and I will try to create a basic version.

**Comment by SimonBaeumer** [comment]
@NissesSenap Repository created: https://github.com/stackrox/operator-helm-chart

**Comment by Maddosaurus** [comment]
Small update on this: In the meantime, we have updated our release process and automatically generate Helm charts that deploy StackRox (OSS) images from quay.io/stackrio-io. You can find these charts [here](https://github.com/stackrox/helm-charts/tree/main/opensource).
I hope this helps building a community operator for StackRox 🎉

**Comment by NissesSenap** [comment]
@SimonBaeumer I won't have time to look in to this the coming weeks. Lots to do during the summer and apparently some PTO as well.
So if anyone have time to look in to this feel free to go for it.

**Comment by NissesSenap** [comment]
Might be worth looking in to: https://github.com/arttor/helmify as an option to help to generate the helm chart for the operator.

**Comment by msugakov** [comment]
Hi @NissesSenap

Could you please clarify what's your ask here?

Are you looking for a way to deploy StackRox via Helm chart or are you looking for a way to deploy StackRox Operator via Helm chart?

The latter would mean to install StackRox Operator (without requiring OpenShift or OLM) which can then install StackRox once you create Central and Secured Cluster Custom Resources (CRs). If that's the case, could you please tell what's the motivation? We've had some internal discussions about something like that but it should be more like an implementation detail of a deployment mechanism rather than a thing important on its own. If you have a use-case specifically for this, it would be good to learn about it.

If your goal is just to install StackRox via Helm charts, this was recently added.
With each new release charts are published to
1. <https://github.com/stackrox/helm-charts/tree/main/opensource>

See also:
1. <https://artifacthub.io/packages/helm/stackrox/stackrox-central-services>
2. <https://artifacthub.io/packages/helm/stackrox/stackrox-secured-cluster-services>

The way you install StackRox with them is described in
1. <https://github.com/stackrox/stackrox/#quick-installation-using-helm>
2. <https://github.com/stackrox/stackrox/#manual-installation-using-helm>

Would these charts work for the automation you have?

**Comment by msugakov** [comment]
@NissesSenap thanks for clarifying! Indeed, the idea of operators is that they should do more than just install/upgrade/uninstall.

I don't see why GitOps is bound to operators though. I believe GitOps with Helm charts can manage resources quite well. Although this leaves you to implement workload monitoring, auto-scaling, backup and recovery, i.e. things that _can_ be solved by the operator _if_ implemented there.
Note that StackRox Helm charts are production ready, tried and tested and are actually used in the current operator implementation. They are robust enough, if you want to give them a try.

As @SimonBaeumer wrote above, deploying the operator on k8s without OLM **is** in the backlog and in our minds. There's no date or priority on this yet, however, so can't tell when the work begins.

**Comment by NissesSenap** [comment]
Okay I don't think we are talking about the same thing will try to explain how I think from a user point of view.
I guess this issue have two main users groups.

* platform admins
* security engineers

As a **platform developer** I want a simple way of deploying stackrox. 
I want to be able to control all settings inside StackRox thorough kubernets CR:s.

As a **security engineer** I want to be able to have specific access that makes it possible for me to configure stackRox through CR:s.
The CR:s will be in a separate repo that only my team will manage. This to make sure that the security team is the only one that can make changes to configuration inside stackRox.

I understand that the possibility to configure everything through the CR is far away but it doesn't hurt to think long ahead. 
Fully understand that this isn't the most prioritized issue and it's great to see that you have fixed a good helm chart. I'm looking forward to giving it a try.

**Comment by msugakov** [comment]
Fair enough.

I think today StackRox Helm charts expose more configuration settings than the Operator does. However these are things that control deployment of StackRox, not the system content that it deals with, like External Integrations or System Policies.

We have had requests to support External Integrations, System Policies and other things which relate to functioning of StackRox (not just its deployment) as Custom Resources. It could turn to be a bit separate from the StackRox Operator itself because a) the Operator will not define these CR:s (a user does that), and b) the Operator may not be interested in these CR:s because they define something that's of interest of the StackRox itself. It's another workstream which may or may not depend on making the Operator a universal way of deployment.

In short, we also have this work on our radar (again without committed priority or date, as of today). Your input is valuable and I linked it from the internal feature request.

**Comment by sahaniarungod** [comment]
Is Stackrox operator available in openshift ?

**Comment by msugakov** [comment]
No, unfortunately StackRox (open source) operator isn't available at the moment. It is slightly more involved than just pushing images to registries under quay.io/stackrox-io and the team hasn't gotten to implement that yet.

**Comment by sahaniarungod** [comment]
> No, unfortunately StackRox (open source) operator isn't available at the moment. It is slightly more involved than just pushing images to registries under quay.io/stackrox-io and the team hasn't gotten to implement that yet.

Is there any official deployment manifest/kustomize except helm for stackrox installation, please suggest.

**Comment by msugakov** [comment]
@sahaniarungod yes, StackRox can be installed with manifests. The procedure is the same as described for the commercial builds:
1. Install Central Services with `roxctl` https://docs.openshift.com/acs/4.0/installing/installing_other/install-central-other.html#install-using-roxctl-other
2. Install Secured Cluster Services with manifests https://docs.openshift.com/acs/4.0/installing/installing_other/install-secured-cluster-other.html#install-sensor-roxctl_install-secured-cluster-other

When running `roxctl central generate` in the first step, make sure to specify `--image-defaults=opensource`, or in interactive mode
> Enter default container images settings (development_build, stackrox.io, rhacs, opensource); it controls repositories from where to download the images, image names and tags format (default: "development_build"): **opensource**

**Comment by kastl-ars** [comment]
From the missing activity on this and the absence of stackrox-operator in the operatorhub I guess there has not been any activity on the topic in the meantime?

Also there are no instructions to install the operator **with OLM**, as this issue seems to mainly focus on an operator **without OLM**. Or did I just miss them?

I tested the installation using helm and can confirm that it works. But it has a manual step (creating the init-bundle) between installation of the `stackrox-central-services` chart and the `stackrox-secured-cluster-services` chart.

Which means using GitOps to deploy it will not work without manual intervention. Or did I get this wrong?

So, in short, the operator would be the right way to go forward and I would really like to see it... :-)

**Comment by msugakov** [comment]
@kastl-ars while I agree that releasing StackRox (community) operator to operatorhub is desirable and nice, I have to also acknowledge that AFAIK there was no progress in that direction.

One important thing is that creating `init-bundle` is required in the operator-based deployment as well. Fundamentally, there needs to be a way to establish trust between Central Services and Secured Cluster Services given that they can be on different clusters. There is/was no universal way of doing this besides making own implementation. In this case init bundles. To sweeten this a bit, the init bundle can be reused between multiple secured clusters and so it is possible to put it in GitOps (securing it carefully as a secret) and provision Secured Clusters. Yes, you still likely won't be able to deploy both Central and Secured Cluster from scratch with one GitOps-y action.

I'll re-post this ticket internally to the Installation team in case they have some new thoughts.

**Comment by kastl-ars** [comment]
Thanks for the reply,  @msugakov!

> Yes, you still likely won't be able to deploy both Central and Secured Cluster from scratch with one GitOps-y action.

Thanks for the clarification. I now this is not the highest priority, so just as an idea.

The init-bundle uses resources available to or in the central pod. If those resources are available to an operator, once the central services are installed, it could fall back to using those resources, instead of requiring an init-bundle.

Of course, this would be a big rewrite of the existing code, so I understand this is not something done in an afternoon. Or something that is high on the agenda.

Out of curiosity, how does the installation via OpenShift ACS work? I thought there is an operator? Asking for this operator to be open sourced might be much to ask, but at least the general idea could be taken from that operator, right?

**Comment by msugakov** [comment]
What I mentioned as `operator-based deployment` was about ACS OpenShift operator. There **is** `init-bundle` as a step, the official docs describe it well: <https://docs.openshift.com/acs/4.4/installing/acs-high-level-overview.html#installing-ocp-operator-steps>.

Creating an community operator for StackRox would be about making the same operator (99% of code is already there, 50% of scripting already exists) available in the community catalog.

**Comment by kastl-ars** [comment]
> What I mentioned as `operator-based deployment` was about ACS OpenShift operator. There **is** `init-bundle` as a step, the official docs describe it well: https://docs.openshift.com/acs/4.4/installing/acs-high-level-overview.html#installing-ocp-operator-steps.

Ah thanks for the info!

> Creating an community operator for StackRox would be about making the same operator (99% of code is already there, 50% of scripting already exists) available in the community catalog.

That would be really nice. And I daresay it would be nice for RedHat folks to see the operator get some testing in the community before they rollout a new version to the openshift operator. (But maybe it is a little more tricky than that...)

**Comment by msugakov** [comment]
This is a good idea to release the community operator first.

**Comment by porridge** [comment]
> AFAIK there was no progress in that direction

There was **some progress** (for example we run operator-e2e-tests on non-OCP platforms to make sure we) but there are no user-visible deliverables so far.

**Comment by porridge** [comment]
FTR, since v4.11 we're shipping a [stackrox-operator helm chart](https://github.com/stackrox/stackrox/blob/master/operator/install/README.md).
